servers/src/main/java/tachyon/master/next/IndexSet.java

```diff
@@ -0,0 +1,190 @@
+import java.util.Set;
+import java.util.HashSet;
+import java.lang.reflect.Field;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.Maps;
+
+/**
+ * A set of objects that can be indexed by fields of the object, the field type must be comparable.
+ * In operations that need field name as the parameter, if the object does not have the field, a
+ * RuntimeException will be thrown.
+ */
+public class IndexSet<T> {
+  private Set<T> mSet = new HashSet<T>();
+  /** Map from fieldName to a map from fieldValue to set of T */
+  private Map<String, Map<Object, Set<T>>> mSetIndexedByFieldValue;
+  /** Map to cache the relation from fieldName to Field which is set accessible */
+  private Map<String, Field> mFields;
+
+  /**
+   * Construct a new IndexSet.
+   *
+   * @param fields the field names to index the set of objects
+   */
+  public IndexSet(String... fields) {
+    mFields = Maps.newHashMap();
+    for (String field : fields) {
+      mFields.put(field, null);
+    }
+
+    mSetIndexedByFieldValue = Maps.newHashMap();
+    for (String field : fields) {
+      mSetIndexedByFieldValue.put(field, new HashMap<Object, Set<T>>());
+    }
+  }
+
+  /**
+   * Add an object to the set.
+   *
+   * @param object the object to add
+   */
+  public void add(T object) {
+    mSet.add(object);
+
+    for (String field : mFields.keySet()) {
+      Map<Object, Set<T>> fieldValueToSet = mSetIndexedByFieldValue.get(field);
+      Object value = getField(object, field);
+      if (fieldValueToSet.containsKey(value)) {
+        fieldValueToSet.get(value).add(object);
+      } else {
+        Set<T> set = new HashSet<T>();
+        set.add(object);
+        fieldValueToSet.put(value, set);
+      }
+    }
+  }
+
+  /**
+   * @return a set of all objects
+   */
+  public Set<T> all() {
+    return mSet;
+  }
+
+  /**
+   * Whether there is an object with the specified field value in the set.
+   *
+   * @param fieldName the field name
+   * @param value the field value
+   * @return true if there is one such object, otherwise false
+   */
+  public boolean contains(String fieldName, Object value) {
+    return !get(fieldName, value).isEmpty();
+  }
+
+  /**
+   * Get a subset of objects with the specified field value.
+   *
+   * @param fieldName the field name
+   * @param value the field value to be satisfied
+   * @return the set of objects or an empty set if no such object exists
+   */
+  public Set<T> get(String fieldName, Object value) {
+    Set<T> set = mSetIndexedByFieldValue.get(fieldName).get(value);
+    return set == null ? new HashSet<T>() : set;
+  }
+
+  /**
+   * Get an object with the specified field value from the set. Assumption is, in the set of
+   * objects, only one object has the field value.
+   *
+   * @param fieldName the field name
+   * @param value the field value
+   * @return the object or null if there is no such object
+   */
+  public T getSingle(String fieldName, Object value) {
```

Comment:
The javadoc is not accurate if I understand it correctly: if there are more than one object selected, you return the first one, or null if selection fails. So why not name this method `getFirst` ?